### PR TITLE
Added simple error handling

### PR DESCRIPTION
**Pull request by FilipKolodziejczyk**
I had trouble with testing that (no errors for several tests in a row). We'll have to observe it over time

### CODE REVIEW

File: src/core/conversation.py [hunk 1]
```diff
@@ -100,7 +111,31 @@ def perform_next_step(self) -> Message:
...
+        error_count = 0
+        while (
+            last_message.role == ConversationRolesInternalEnum.CODE
+            and "Traceback" in last_message.content
+        ):
+            # Ask code assistant to fix the error and overwrite previous output. Do until error is fixed.
+
+            error_count += 1
+            if error_count >= error_limit:
+                raise Exception("Too many errors in code.")
```
Comment (jswistak) [inline src/core/conversation.py:125]: IMO we should be defining our own exception instead on using default one

File: src/core/conversation.py [hunk 2]
```diff
@@ -100,7 +111,31 @@ def perform_next_step(self) -> Message:
         else:
             raise Exception(f"Invalid conversation role: {last_message.role}")
 
-        return self._get_last_message()
+        last_message = self._get_last_message()
+        error_limit = 5
```
Comment (jswistak) [inline src/core/conversation.py:115]: Why don't we deduce the errors here from conversation messages? Currently we may generate incorrect code all the time and we will make 2,5x more API calls

File: src/core/conversation.py [hunk 3]
```diff
@@ -100,7 +111,31 @@ def perform_next_step(self) -> Message:
...
+                role=ConversationRolesInternalEnum.ANALYSIS,
+                content="Execution of provided code failed. Please fix the error and try again. \n\nHere is the error message:\n```"
+                + last_message.content
+                + "```",
+            )
+            self._send_message_code()
+            # Remove artificial message and previous code output
+            self._conversation.pop(-2)
+            self._conversation.pop(-2)
+            last_message = self._get_last_message()
```
Comment (jswistak) [inline src/core/conversation.py:136]: Good idea, but it may cause conversation to behave in a strange way where the context of the message may be lost.

**Review by jswistak — CHANGES_REQUESTED** [review]

File: src/core/conversation.py (hunk 1 above)
Comment (FilipKolodziejczyk) [reply src/core/conversation.py:125]: Okay, will do

File: src/core/conversation.py (hunk 2 above)
Comment (FilipKolodziejczyk) [reply src/core/conversation.py:115]: I will reduce limit to 3

File: src/core/conversation.py [hunk 4]
```diff
@@ -1,12 +1,18 @@
...
 from models.models import ConversationRolesInternalEnum, LLMType, Message
 from prompt_manager.ipromptmanager import IPromptManager
 from runtime.iruntime import IRuntime
 
 
+class CodeRetryLimitExceeded(Exception):
+    """Exception raised when too many errors occur during code execution."""
+
+    pass
+
```
Comment (jswistak) [inline src/core/conversation.py:14]: Not sure if it will work or if the __init__ will be needed

File: src/core/conversation.py [hunk 5]
```diff
@@ -100,7 +117,31 @@ def perform_next_step(self) -> Message:
...
+        error_count = 0
+        while (
+            last_message.role == ConversationRolesInternalEnum.CODE
+            and "Traceback" in last_message.content
+        ):
+            # Ask code assistant to fix the error and overwrite previous output. Do until error is fixed.
+
+            error_count += 1
+            if error_count >= error_limit:
+                raise CodeRetryLimitExceeded()
```
Comment (jswistak) [inline src/core/conversation.py:131]: some message should be passed

**Review by jswistak — CHANGES_REQUESTED** [review]

**Review by jswistak — APPROVED** [review]

**Pull request merged**